Wayfinding pinpointers for Kilo, Donut, Delta, Box and Meta

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds ~~invisible, indestructible~~ wayfinding beacons. These are children of bot navigation beacons. They go in common locations to help guide crew.

Adds wayfinding pinpointers. These point to navigation beacons with "wayfinding" set in their codes list. They use the wayfinding value as the name that shows up on pinpointers. e.g. wayfinding = Head of Personnel or wayfinding = coom if a player edits it.

Adds a wayfinding pinpointer dispenser. This goes at arrivals and dispenses one pinpointer per name.

This is meant to help new players, players on new maps, and players on lesser played maps find their way around.

This pull also fixes an issue with navigation beacons that may not have been reported yet. I'm assuming changing them to use global lists caused it.

I didn't know this until this pull, but with tools and access you can modify a navigation beacon. You can change a MULEbot delivery beacon to a bot pathfinding beacon, vice versa, or to both. You can also change a beacon's location variable, delivery direction, or the next beacon in the bot patrol path. This pull makes it so you can change whether a beacon works as a wayfinding beacon and the name it gives to pinpointers too.

The aforementioned issue was when players modified beacons it didn't update their global lists. So you couldn't change a beacon's function or location name. Beacons update these lists now.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

I've heard people express frustration finding their way around these maps as well as experienced it myself. I think one of the reasons Meta is preferred is because more people know the layout.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
add: Struggling to find your way around Kilo, Delta, Donut or even Box and Meta? Look for the pinpointer dispenser at arrivals.
fix: Modifying a navigation beacon's codes list or location name now correctly updates the associated global lists (this is an IC thing, try it)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

optimumtact commented:
`invisible, indestructible beacons`

why, these can easily be built into the departments and interactable, giving people another interesting thing to play with.

wesoda25 commented:
It'd be cool if it spawned in new players pockets if their hours are under 20 or something

cacogen commented:
@optimumtact 
I can make them part of the world. What interactions do you have in mind?

cacogen commented:
These are children of existing navigation beacons now, which means their name and purpose can be tinkered with.

bobbahbrown commented on code/game/objects/items/wayfinding.dm:
```diff
@@ -0,0 +1,172 @@
+	var/list/obj/item/pinpointer/wayfinding/pinpointers = list()
+
+/obj/machinery/pinpointer_dispenser/attack_hand(mob/living/carbon/user)
+	if((user.name in pinpointers))
```
```suggestion
	if(user.name in pinpointers)
```

MrDoomBringer commented:
@cacogen also maybe add the waypoint beacon's locations to the station maps that exist on kilo, etc.?

cacogen commented:
just had a round on donut 
>[Common] Lou Parkinson says, "Where the fuck is the shuttle"

>[Common] Hercules Peachmangler asks, "Where is EVAC?"

>Alice Smith asks, "WHERE IS SHUTTLE?"

>(F) Sild Solace whispers, "And I never found EVA on this shitbox station... who the fuck thinks a ring structure that isn't spinning in space is a good idea?"

i'm gonna add wayfinding beacons to that map and add a beacon for EVA as well

@MrDoomBringer
i'll look at that

@wesoda25 
i'll figure out how to do that

cacogen commented on code/game/objects/items/wayfinding.dm:
```diff
@@ -0,0 +1,112 @@
+/obj/machinery/navbeacon/wayfinding
+	name = "wayfinding beacon"
+	desc = "A beacon used by wayfinding pinpointers."
+//	var/id
```
me, get rid of this

BadSS13Player commented:
Why not have them on Metastation too, for completeness' sake?

cacogen commented:
laziness and lack of faith in the popularity of the feature combined with meta's ubiquitousness meaning regulars know the layout unlike maps like kilo or donut or even box nowadays
i'm not against the idea, in theory every map should have them if maintainers/oranges accept this is a valid problem and this is a workable solution

SpaceManiac commented:
I approve of the basic design and the bugfix

cacogen committed:
Wayfinding pinpointers for Kilo, Donut, Delta and Box

cacogen committed:
Navigationally Challenged quirk

Spawn with wayfinder. Automatically given to clients less than 7 days old.

cacogen committed:
Adds beacons to Meta

- Wayfinding beacons now try to grab their location from any airlock they're under first, then their area, if it hasn't been set on the map

cacogen commented:
@BadSS13Player

Jalleo commented:
Related question. Can the pinpointers tell you when you need to go up or down a z?
Secondly this is a great PR and idea. I heavily approve. Maybe we could add a way to construct these and the other beacons.

cacogen commented:
As with all pinpointers (to my knowledge at least) if you're off z-level it shows a question mark

kriskog commented:
Merge conflict

ExcessiveUseOfCobblestone commented on code/controllers/subsystem/job.dm:
```diff
@@ -473,6 +473,8 @@ SUBSYSTEM_DEF(job)
 	if(ishuman(living_mob))
 		var/mob/living/carbon/human/wageslave = living_mob
 		living_mob.add_memory("Your account ID is [wageslave.account_id].")
+		if(job.give_wayfinder(M.client) && !living_mob.has_quirk(/datum/quirk/needswayfinder))
```
why make a proc for this? just do the check here /w comment

```suggestion
		if(M.client.player_age < 7 && !living_mob.has_quirk(/datum/quirk/needswayfinder))
```

cacogen commented:
>commits 127

all in a day's work

cacogen committed:
Removes unnecesssary proc

Co-Authored-By: moo <11748095+ExcessiveUseOfCobblestone@users.noreply.github.com>

kriskog commented on code/game/objects/items/wayfinding.dm:
```diff
@@ -0,0 +1,77 @@
+/obj/item/pinpointer/wayfinding/attack_self(mob/living/user)
+	if(active)
+		toggle_on()
+		user.visible_message("<span class='notice'>[user] deactivates [user.p_their()] pinpointer.</span>", "<span class='notice'>You deactivate your pinpointer.</span>")
```
This wont show the user anything. Only blind_message shows when the object is inside something, in this case inside user. Use a to_chat instead, the people around him also dont need this feedback.

kriskog commented on code/game/objects/items/wayfinding.dm:
```diff
@@ -0,0 +1,77 @@
+		beacons[B.codes["wayfinding"]] = B
+
+	if(!beacons.len)
+		user.visible_message("<span class='notice'>[user]'s pinpointer fails to detect a signal.</span>", "<span class='notice'>Your pinpointer fails to detect a signal.</span>")
```
Another to_chat

kriskog commented on code/game/objects/items/wayfinding.dm:
```diff
@@ -0,0 +1,77 @@
+
+	target = beacons[A]
+	toggle_on()
+	user.visible_message("<span class='notice'>[user] activates [user.p_their()] pinpointer.</span>", "<span class='notice'>You activate your pinpointer.</span>")
```
and another

kriskog commented on code/modules/jobs/job_types/_job.dm:
```diff
@@ -142,7 +142,6 @@
 		return TRUE	//Available in 0 days = available right now = player is old enough to play.
 	return FALSE
 
-
```
Only change in this file

cacogen commented on code/game/objects/items/wayfinding.dm:
```diff
@@ -0,0 +1,77 @@
+		owner = user.name
+	else if(owner != user.name)
+		to_chat(user, "<span class='notice'>The pinpointer doesn't respond. It seems to only recognise its owner.</span>")
+		return
```
Disincentivise theft if we're limiting it to one pinpointer per user

cacogen commented on code/game/objects/items/wayfinding.dm:
```diff
@@ -0,0 +1,77 @@
+	var/obj/item/pinpointer/wayfinding/P = new /obj/item/pinpointer/wayfinding(get_turf(src))
+	user.put_in_hands(P)
+	P.owner = user.name
+	pinpointers[user.name] = P
```
Thought this might be a workaround for that 1% of rounds you have two people spawn with the same name (might be more actually considering genetics and changelings)

kriskog replied:
That would be fine for the outward facing stuff of the pinpointer, or grabbing the name from your ID and restrictkngkt to people with one. The issue is this one is added to a list you check to decide if they can have another, and two unknowns would interfere with each other.

cacogen commented on code/controllers/subsystem/job.dm:
```diff
@@ -473,6 +473,8 @@ SUBSYSTEM_DEF(job)
 	if(ishuman(living_mob))
 		var/mob/living/carbon/human/wageslave = living_mob
 		living_mob.add_memory("Your account ID is [wageslave.account_id].")
+		if(living_mob.client.player_age < 7 && !living_mob.has_quirk(/datum/quirk/needswayfinder))
```
thanks!

kriskog commented:
No pubby changes by the way, not doing that one?

cacogen committed:
kriskog changes

cacogen commented:
Yeah probably not
